[ASCellNode] Added indexPath property.

george-gw opened the pull request:
refs #2447 

[ASCellNode]:
- Added indexPath property.
- Added a test case.

maicki commented on AsyncDisplayKit/ASCellNode.h:
```diff
@@ -108,6 +108,14 @@ typedef NS_ENUM(NSUInteger, ASCellNodeVisibilityEvent) {
+ *
+ * @note This property must be accessed on the main thread.
+ */
+@property (nonatomic, readonly, nullable) NSIndexPath *indexPath;
```
Please add `strong`

maicki commented on AsyncDisplayKit/ASCellNode.mm:
```diff
@@ -204,6 +204,26 @@ - (void)__setHighlightedFromUIKit:(BOOL)highlighted;
+  UIScrollView *scrollView = self.scrollView;
+
+  ASDisplayNode *owningNode = scrollView.asyncdisplaykit_node;
+  if ([owningNode isKindOfClass:[ASCollectionNode class]]) {
```
The isKindOfClass: calls are pretty heavy ... can we move this logic in setScrollView: and cache via a flags property? Look into `ASCollectionView` / `ASTableView` how we check if delegate methods are implemented.

george-gw replied:
Done

Adlai-Holler commented on AsyncDisplayKit/ASCellNode.mm:
```diff
@@ -204,6 +244,26 @@ - (void)__setHighlightedFromUIKit:(BOOL)highlighted;
+  } else if (_scrollViewType.isTableNode) {
+    return [(ASTableNode *)owningNode indexPathForNode:self];
+  } else if (_scrollViewType.isCollectionView) {
+    return [(ASCollectionView *)owningNode indexPathForNode:self];
```
`scrollView indexPathForNode:`

That said, I think we should actually remove support for ASCollectionView and ASTableView from here, and only work with the ASTableNode and ASCollectionNode. 

The rationale here is,
- The implementation is much simpler.
- The views return index paths in the view-space, which are pretty useless to developers.

george-gw replied:
`scrollView` would return ASCollectionView or ASTableView, we need the owningNode which is the corresponding ASCollectionNode or ASTableNode.

Aside from that I removed support for ASCollectionView and ASTableView as requested.

Adlai-Holler commented on AsyncDisplayKit/ASCellNode.h:
```diff
@@ -108,6 +108,14 @@ typedef NS_ENUM(NSUInteger, ASCellNodeVisibilityEvent) {
+ *
+ * @note This property must be accessed on the main thread.
+ */
+@property (nonatomic, strong, readonly, nullable) NSIndexPath *indexPath;
```
No need for `strong` here. Computed properties having no storage qualifier is a pattern e.g. `UICollectionView.visibleCells`.

maicki commented on AsyncDisplayKit/ASCellNode.mm:
```diff
@@ -186,6 +194,38 @@ - (void)setHighlighted:(BOOL)highlighted
+    _scrollViewType.isTableNode      = 0;
+    _scrollViewType.isCollectionView = 0;
+    _scrollViewType.isCollectionNode = 0;
+  }
```
No need for setting the flags one after another to `0`. Just add at the beginning of the if else block:

```Objc
memset(&_scrollViewType, 0, sizeof(_scrollViewType));
```

 and in the if else set the once to `1`. You can than also remove the else block completely

Adlai-Holler commented:
The CI failure is spurious and can be ignored.

george-gw commented:
A note: `scrollView`is only non-nil when the cell is displayed:

```
- (void)collectionView:(UICollectionView *)collectionView willDisplayCell:(_ASCollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath
{
  ASCellNode *cellNode = [cell node];
  cellNode.scrollView = collectionView;
```

```
- (void)collectionView:(UICollectionView *)collectionView didEndDisplayingCell:(_ASCollectionViewCell *)cell forItemAtIndexPath:(NSIndexPath *)indexPath
{
  ASCellNode *cellNode = [cell node];
  [...]
  cellNode.scrollView = nil;
```

Should this case be handled in some way? If so, how would you recommend doing that?

Adlai-Holler commented:
@george-gw You're absolutely right. 

One solution would be to add a weak reference to the owning `ASTableNode/ASCollectionNode` to `ASIndexedNodeContext` (passed into `init`) and to `ASCellNode` (publicly `readonly`, privately `readwrite` in `ASCellNode+Internal.h`). When `ASIndexedNodeContext` creates the node, we can attach that pointer to the cell node. 

I think that should work pretty nicely. Thoughts?

george-gw commented:
@Adlai-Holler I changed the code to better fit our needs based on your suggestions, I ran all the test cases and they all succeeded, I added a few more checks in the indexPath consistency test to check for nodes that aren't displayed and it seems to work fine.

Adlai-Holler commented:
test this please

george-gw commented:
@Adlai-Holler There is a test case for indexPath consistency, it tests different aspects. Do you have a different kind of test you'd like me to write?

Adlai-Holler commented:
@george-gw Oh those are just the magic words that you use to get our Jenkins CI to run tests. Your test looks good.

george-gw commented:
@Adlai-Holler oh haha! 👍

Adlai-Holler commented:
@george-gw Take a look at #2493. Now that `ASCellNode` has `indexPath`, that custom subclass will not automatically synthesize an ivar for it. Could you modify the example code to get Jenkins passing?

george-gw commented:
@Adlai-Holler which would you prefer? changing the property name to `currentIndexPath` or just doing something like: `@synthesize indexPath=__indexPath;`?

Adlai-Holler commented:
add to whitelist and test this please

Adlai-Holler commented:
Thanks @george-gw. Resolves #2447. Here, we, go!

hannahmbanana commented:
@Adlai-Holler @george-gw: should we rename the `indexPath` property to something more unique? This currently conflicts with Pinterest code.
